app/controllers/tag_implication_controller.rb

```diff
@@ -1,76 +1,96 @@
-        redirect_to :controller => "job_task", :action => "index"
-      else
-        access_denied
-      end
+    when 'Delete'
+      delete_tags(ids)
+    when 'Approve'
+      approve_tags(ids)
     else
       head :bad_request
     end
   end
 
   def index
-    if params[:commit] == "Search Aliases"
-      return redirect_to :controller => "tag_alias", :action => "index", :query => params[:query]
-    end
+    return redirect_to_tag_alias_index if searching_aliases?
+
+    @implications = TagImplication.retrieve_matching_tag_id
+    @implications = TagImplication.apply_query_filter(params, @implications)
+    @implications = @implications.paginate(page: page_number, per_page: 20)
+    respond_to_list('implications')
+  end
+
+  private
+
+  def redirect_to_tag_alias_index
+    redirect_to(controller: 'tag_alias', action: 'index', query: params[:query])
+  end
+
+  def searching_aliases?
+    params[:commit] == 'Search Aliases'
+  end
 
-    # FIXME: subquery in order
-    @implications = TagImplication.order(Arel.sql("is_pending DESC, (SELECT name FROM tags WHERE id = tag_implications.predicate_id), (SELECT name FROM tags WHERE id = tag_implications.consequent_id)"))
+  def approve_tags(ids)
+    if @current_user.is_mod_or_higher?
+      ids.each do |x|
+        approve_tag(x)
+      end
 
-    if params[:query]
-      tag_ids = Tag.where("name ILIKE ?", "*#{params[:query]}*".to_escaped_for_sql_like).select(:id)
-      @implications = @implications
-        .where("predicate_id IN (?) OR consequent_id IN (?)", tag_ids, tag_ids)
-        .order(:is_pending => :desc, :consequent_id => :asc)
+      flash[:notice] = 'Tag implication approval jobs created'
+      redirect_to controller: 'job_task', action: 'index'
+    else
+      access_denied
     end
+  end
 
-    @implications = @implications.paginate :page => page_number, :per_page => 20
+  def approve_tag(x)
+    if CONFIG['enable_asynchronous_tasks']
+      JobTask.create(task_type: 'approve_tag_implication', status: 'pending', data: format_with_user_info(x))
+    else
+      find_and_approve(x)
+    end
+  end
+
+  def delete_tags(ids)
+    if @current_user.is_mod_or_higher? || check_creator_tag(ids)
+      ids.each { |x| destroy_tag_and_notify_user(x) }
 
-    respond_to_list("implications")
+      flash[:notice] = 'Tag implications deleted'
+      redirect_to action: 'index'
+    else
+      access_denied
+    end
   end
 
-  private
+  def destroy_tag_and_notify_user(x)
+    TagImplication.find(x).destroy_and_notify(@current_user, params[:reason])
+  end
 
   def tag_implication_params
     params.require(:tag_implication).permit(:predicate, :consequent, :reason)
   end
+
+  def check_creator_tag(ids)
+    TagImplication.where(id: ids, is_pending: true, creator_id: @current_user.id).count == ids.count
+  end
+
+  def format_with_user_info(x)
+    { id: x, updater_id: @current_user.id, updater_ip_addr: request.remote_ip }
```

Comment:
as mentioned before, this doesn't match with the format of the actual data. The keys are string.

Reply:
You're correct. It should have been { 'id' => x, ... } and so on